### PR TITLE
Add OGN buyback contract

### PR DESCRIPTION
**Pull request by tomlinton**
Deploys a simple buyback contract that swaps OUSD for OGN via Uniswap. The buyback contract replaces the trustee and the swap function is called as part of allocate. Some design decisions I made:

- Non upgradeable, has a hardcoded Uniswap path and hardcoded threshold for swap amounts
- Only swap if buyback has > 1000 OUSD

Contract change checklist:
  - [ ] Code reviewed by 2 reviewers. See [template](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review.md) and [example](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review-Example.md).
  - [x] Unit tests pass
  - [x] Slither tests pass with no warning
  - [ ] Echidna tests pass if PR includes changes to OUSD contract (not automated, run manually on local)

### CODE REVIEW

File: contracts/contracts/buyback/Buyback.sol [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        require(vaultAddr == msg.sender, "Caller is not the Vault");
+        _;
+    }
+
+    /**
+     * @dev Set address of Uniswap for performing liquidation of strategy reward
+     * tokens. Setting to 0x0 will pause swaps.
+     * @param _address Address of Uniswap
+     */
+    function setUniswapAddr(address _address) external onlyGovernor {
```
Comment (franckc) [inline contracts/contracts/buyback/Buyback.sol:42]: should this be onlyStrategist so that the strategist can pause swaps?

File: contracts/contracts/buyback/Buyback.sol [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    /**
+     * @dev Execute a swap of OGN for OUSD via Uniswap or Uniswap compatible
+     * protocol (e.g. Sushiswap)
+     **/
+    function swap() external onlyVault {
+        if (uniswapAddr != address(0)) {
+            uint256 sourceAmount = ousd.balanceOf(address(this));
+            if (sourceAmount >= 1000 * 1e18) {
```
Comment (franckc) [inline contracts/contracts/buyback/Buyback.sol:57]: should this be a threshold stored as a public variable with a setter accessible by the strategist?

File: contracts/contracts/buyback/Buyback.sol [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                // Uniswap redemption path
+                address[] memory path = new address[](4);
+                path[0] = address(ousd);
+                path[1] = address(usdt); // USDT
+                path[2] = IUniswapV2Router(uniswapAddr).WETH();
+                path[3] = address(ogn);
+                IUniswapV2Router(uniswapAddr).swapExactTokensForTokens(
```
Comment (franckc) [inline contracts/contracts/buyback/Buyback.sol:64]: Should we emit a BuyBack event for our analytics needs?
Or is it pretty stratight forward to leverage the events this method I'm sure already emits?

File: contracts/deploy/015_buyback.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+      args: [dVaultCore.address],
+    },
+    {
+      contract: cBuyback,
+      signature: "claimGovernance()",
+    },
+    {
+      contract: cVaultAdmin,
+      signature: "setTrusteeAddress(address)",
+      args: [cBuyback.address],
```
Comment (franckc) [inline contracts/deploy/015_buyback.js:65]: Should this be done separately, after the governor claimed governance?

**Review by franckc — COMMENTED** [review]
Sorry sorry for being super later on this review.

A few minor questions inline

**Review by DanielVF — APPROVED** [review]
I still need to set up to run a fork test and verify that the uniswap calls work correctly.
#### Requirements

I'm wondering if we should just keeper this directly and keep the vault unmodified? I guess doing it this way adds gas cost, pushing back profitability on slippage attacks a little, but at the cost of us paying the gas all the time.

#### Internal State

_No stored state :heart:_

#### Attack

The obvious attack here is a slippage attack on the sale of the tokens. OUSD/USDT has fairly thin liquidity, with 2.3 million on each side.

If my math is right, for roughly 300 dollars in Uniswap fees, an attacker could bend Uniswap, then call this and capture 5% of the sale value. That would make this profitable, ignoring gas, at a trade of $6,000 of yield. There is currently $3,941 OUSD ready to be swapped. Now gas adds a little to this, and we should be running this more often once it's automated. But we are inside the order of magnitude where these attacks make sense.

Having this called only by the vault doesn't move running this action out of user control, just adds a little gas

#### Logic

_Logic good._

#### Tests

  - [x] Each logical branch has a test
  - [x] Edge conditions are tested

#### Flavor

See inline comments.

#### Overflow

No overflow, though a probably unnecessary plain addition is in the code and can be removed.

#### Black magic

_No black magic._

#### Authentication

- [x] Never use tx.origin for authentication
- [x] Check that every external/public function should actually be external
- [x] Check that every external/public function has the correct authentication

#### Cryptographic code

_No cryptographic code_

#### Gas problems

_No gas issues, but see inline comments on gas savings._

#### External calls

- [x] Contract addresses passed in are validated
- [x] Unsafe external calls
- [x] Reentrancy guards on all state changing functions
- [-] How vulnerable is this to a slippage attack
- [-] Does this work on the fork (STILL NEED TO TEST)

#### Ethereum

- [x] Contract does not send or receive Ethereum.
- [x] Contract has no payable methods.

File: contracts/contracts/buyback/Buyback.sol (hunk 3 above)
Comment (DanielVF) [reply contracts/contracts/buyback/Buyback.sol:64]: For the flipper I just planned on watching the ERC20 Transactions out of it.

File: contracts/contracts/buyback/Buyback.sol [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+     * protocol (e.g. Sushiswap)
+     **/
+    function swap() external onlyVault {
+        if (uniswapAddr != address(0)) {
+            uint256 sourceAmount = ousd.balanceOf(address(this));
+            if (sourceAmount >= 1000 * 1e18) {
+                // Uniswap redemption path
+                address[] memory path = new address[](4);
+                path[0] = address(ousd);
+                path[1] = address(usdt); // USDT
```
[the remark this answers is not in the document]
Comment (tomlinton) [reply contracts/contracts/buyback/Buyback.sol:61]: Yep thanks will remove, the address was hardcoded here before but I moved it up to the top and didn't remove that comment.

File: contracts/deploy/015_buyback.js (hunk 4 above)
Comment (tomlinton) [reply contracts/deploy/015_buyback.js:65]: I don't see a reason to, the whole thing either works or it doesn't, any reason you had in mind?

File: contracts/contracts/buyback/Buyback.sol (hunk 2 above)
Comment (tomlinton) [reply contracts/contracts/buyback/Buyback.sol:57]: I'm ok with it being hardcoded. I don't think we will need to change it and we could also just redeploy it if we did. Happy to hear any argument for making it configurable though.

File: contracts/contracts/buyback/Buyback.sol (hunk 3 above)
Comment (franckc) [reply contracts/contracts/buyback/Buyback.sol:64]: Sounds good.

**Review by franckc — APPROVED** [review]

File: contracts/contracts/buyback/Buyback.sol [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                path[2] = IUniswapV2Router(uniswapAddr).WETH();
+                path[3] = address(ogn);
+                IUniswapV2Router(uniswapAddr).swapExactTokensForTokens(
+                    sourceAmount,
+                    uint256(0),
+                    path,
+                    address(this),
+                    now + 1800
```
[the remark this answers is not in the document]
Comment (tomlinton) [reply contracts/contracts/buyback/Buyback.sol]: Will change

File: contracts/contracts/buyback/Buyback.sol [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        ousd.safeApprove(uniswapAddr, uint256(-1));
+        emit UniswapUpdated(_address);
+    }
+
+    /**
+     * @dev Execute a swap of OGN for OUSD via Uniswap or Uniswap compatible
+     * protocol (e.g. Sushiswap)
+     **/
+    function swap() external onlyVault {
+        if (uniswapAddr != address(0)) {
```
[the remark this answers is not in the document]
Comment (tomlinton) [reply contracts/contracts/buyback/Buyback.sol]: Changed 👍

File: contracts/contracts/buyback/BuybackConstructor.sol [hunk 8]
```diff
@@ -0,0 +1,25 @@
+pragma solidity 0.5.11;
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+import { Buyback } from "./Buyback.sol";
+
+contract BuybackConstructor is Buyback {
```
[the remark this answers is not in the document]
Comment (tomlinton) [reply contracts/contracts/buyback/BuybackConstructor.sol:7]: Yea I was looking for a way to decrease gas consumption of the production contract through constants but also have it configurable for tests. I didn't fully follow through here as you've spotted. I looked at the @eth-optimisim/smock package as an option.

I could duplicate the code like Flipper/FlipperDev, but surely there must be an easier way. I don't like ending up with two separate contracts, one for tests and one for the deployment.

[the remark this answers is not in the document]
Comment (tomlinton) [reply contracts/contracts/buyback/BuybackConstructor.sol:7]: I guess one easy solution is just to change the constants at deploy time

[the remark this answers is not in the document]
Comment (franckc) [reply contracts/contracts/buyback/BuybackConstructor.sol:7]: Agreed it is not the prettiest, but for now in the spirit on moving fast and focusing on NFT, should we go for the simplest solution which is to duplicate the contract? e.g. BuyBack.sol and BuyBackDev.sol
We can file an issue for figuring our a better solution later?

**Comment by franckc** [comment]
FYI - I'm starting to gather the list of contracts to include when we do our next mainnet deploy so that I can iron out all the deploy scripts involved. Trying to target end of this week.

**Comment by franckc** [comment]
Ah also, there is a test failure which seems related to this PR. Let's fix it before merging.

**Comment by tomlinton** [comment]
I fixed the slither error. Not sure about that timeout, but seems to be unrelated to this. I'll keep pressing retry till it goes green!

File: contracts/contracts/buyback/BuybackConstructor.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (tomlinton) [reply contracts/contracts/buyback/BuybackConstructor.sol]: Went with the duplicate and changed the prod to constant similar to Flipper.
